PerformanceMonitor.bsv

```diff
@@ -0,0 +1,163 @@
+ *   http://www.beri-open-systems.org/legal/license-1-0.txt
+ *
+ * Unless required by applicable law or agreed to in writing, Work distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * @BERI_LICENSE_HEADER_END@
+ */
+
+package PerformanceMonitor; 
+
+import Vector :: * ;
+
+// A module which wants to count events must return some vector of events
+// However it is easier to count the events as a Struct within a module
+// So if the Struct is an instance of this class it can be returned as is
+typeclass BitVectorable #(type from, numeric type n, numeric type m) dependencies (from determines (n, m));
+  function Vector #(m, Bit #(n)) to_vector (from e);
+endtypeclass
+
+instance BitVectorable #(Vector #(m, b), nn, mm)
+   provisos (Bits #(b, n), Add #(a__, n, nn), Add #(b__, m, mm));
+   function Vector #(mm, Bit #(nn)) to_vector (Vector #(m, b) e);
+      return append (map (zeroExtend, map (pack, e)), replicate (0));
+   endfunction
+endinstance
+
+function Vector #(mm, Bit #(nn)) to_large_vector (from e)
+   provisos (BitVectorable #(from, n, m), Add #(a__, n, nn), Add #(b__, m, mm));
+   return append (map (zeroExtend, to_vector (e)), replicate (0));
+endfunction
+
+// Write is exposed to only one counter per cycle
+// Could change write_* methods to return WriteOnly Vector if needed
+interface PerfCounters_IFC #(numeric type ctrs, numeric type ctrW, numeric type evts);   
+  (* always_ready, always_enabled *)
+  method Action send_performance_events (Vector #(evts, Bit #(ctrW)) evts);
+
+  (* always_ready *)
+  method Vector #(ctrs, ReadOnly #(Bit #(ctrW))) read_counters;
+  (* always_ready *)
+  method Vector #(ctrs, ReadOnly #(Bit #(TLog #(evts)))) read_ctr_sels;
+  (* always_ready *)
+  method ReadOnly #(Bit #(ctrs)) read_ctr_inhibit;
+  (* always_ready *)
+  method ReadOnly #(Bit #(ctrs)) read_ctr_overflow;
+
+  (* always_ready *)
+  method Action write_counter (Bit #(TLog #(ctrs)) idx, Bit #(ctrW) val);
+  (* always_ready *)
+  method Action write_ctr_sel (Bit #(TLog #(ctrs)) idx, Bit #(TLog #(evts)) val);
+  (* always_ready *)
+  method Action write_ctr_inhibit (Bit #(ctrs) val);
+endinterface
+
+
+module mkPerfCounters_Core (PerfCounters_IFC #(ctrs, ctrW, evts));
+  let ctrs = valueOf(ctrs);
+  let ctrW = valueOf(ctrW);
+  let evts = valueOf(evts);
+
+  // Event counters
+  Vector #(ctrs, Array #(Reg #(Bit #(ctrW))))
+    vec_rg_counter <- replicateM(mkCRegU(2));
+  // Which event the corresponding counter will count
+  Vector #(ctrs, Reg #(Bit #(TLog #(evts))))
+    vec_rg_event_sel <- replicateM(mkRegU);
+
+  Reg #(Bit #(ctrs)) rg_ctr_inhibit <- mkReg(0);
+  Wire #(Bit #(ctrs)) wr_overflow <- mkDWire(0);
+
+  // Since [i] is not a function, cannot map it to a Vector
+  function a_type sel_snd_port (Array# (a_type) arr) = arr[1];
```

Comment:
```suggestion
  function a_type sel_snd_port (Array #(a_type) arr) = arr[1];
```

Reply:
Though this is BlueStuff so you should probably follow its code style and drop the space altogether. You're also inconsistent throughout about whether it's `fn (` or `fn(`, with the latter being the BlueStuff style.

Reply:
Should be fixed everywhere now